mongo/client_bulk_write.go

```diff
@@ -0,0 +1,703 @@
+				multi:          true,
+				checkDollarKey: true,
+			}).marshal(mb.client.bsonOpts, mb.client.registry)
+		case *ClientReplaceOneModel:
+			mb.cursorHandlers = append(mb.cursorHandlers, mb.appendUpdateResult)
+			doc, err = (&clientUpdateDoc{
+				namespace:      nsIdx,
+				filter:         model.Filter,
+				update:         model.Replacement,
+				hint:           model.Hint,
+				arrayFilters:   nil,
+				collation:      model.Collation,
+				upsert:         model.Upsert,
+				multi:          false,
+				checkDollarKey: false,
+			}).marshal(mb.client.bsonOpts, mb.client.registry)
+		case *ClientDeleteOneModel:
+			mb.cursorHandlers = append(mb.cursorHandlers, mb.appendDeleteResult)
+			doc, err = (&clientDeleteDoc{
+				namespace: nsIdx,
+				filter:    model.Filter,
+				collation: model.Collation,
+				hint:      model.Hint,
+				multi:     false,
+			}).marshal(mb.client.bsonOpts, mb.client.registry)
+		case *ClientDeleteManyModel:
+			canRetry = false
+			mb.cursorHandlers = append(mb.cursorHandlers, mb.appendDeleteResult)
+			doc, err = (&clientDeleteDoc{
+				namespace: nsIdx,
+				filter:    model.Filter,
+				collation: model.Collation,
+				hint:      model.Hint,
+				multi:     true,
+			}).marshal(mb.client.bsonOpts, mb.client.registry)
+		default:
+			mb.cursorHandlers = append(mb.cursorHandlers, nil)
+		}
+		if err != nil {
+			return 0, nil, err
+		}
+		length := len(doc)
+		if !exists {
+			length += len(ns)
+		}
+		size += length
+		if size >= totalSize {
+			break
+		}
+
+		dst = fn.appendDocument(dst, strconv.Itoa(n), doc)
+		if !exists {
+			idx, doc := bsoncore.AppendDocumentStart(nil)
+			doc = bsoncore.AppendStringElement(doc, "ns", ns)
+			doc, _ = bsoncore.AppendDocumentEnd(doc, idx)
+			nsDst = fn.appendDocument(nsDst, strconv.Itoa(n), doc)
+		}
+		n++
+	}
+	if n == 0 {
+		return 0, dst[:l], nil
+	}
+
+	dst = fn.updateLength(dst, opsIdx, int32(len(dst[opsIdx:])))
+	nsDst = fn.updateLength(nsDst, nsIdx, int32(len(nsDst[nsIdx:])))
+	dst = append(dst, nsDst...)
+
+	mb.retryMode = driver.RetryNone
+	if mb.client.retryWrites && canRetry {
+		mb.retryMode = driver.RetryOnce
+	}
+	return n, dst, nil
+}
+
+func (mb *modelBatches) processResponse(ctx context.Context, resp bsoncore.Document, info driver.ResponseInfo) error {
+	var writeCmdErr driver.WriteCommandError
+	if errors.As(info.Error, &writeCmdErr) && writeCmdErr.WriteConcernError != nil {
+		wce := convertDriverWriteConcernError(writeCmdErr.WriteConcernError)
+		if wce != nil {
+			mb.writeConcernErrors = append(mb.writeConcernErrors, *wce)
+		}
+	}
+	if len(resp) == 0 {
+		return nil
+	}
+	var res struct {
+		Ok        bool
+		Cursor    bsoncore.Document
+		NDeleted  int32
+		NInserted int32
+		NMatched  int32
+		NModified int32
+		NUpserted int32
+		NErrors   int32
+		Code      int32
+		Errmsg    string
+	}
+	dec := bson.NewDecoder(bson.NewDocumentReader(bytes.NewReader(resp)))
+	dec.SetRegistry(mb.client.registry)
+	err := dec.Decode(&res)
```

Comment:
Using the user-defined registry is not necessary for decoding command documents (only for encoding/decoding user document data).

```suggestion
	err := bson.Unmarshal(resp, &res)
```